src/core/base64.cpp

```diff
@@ -54,4 +54,134 @@ QVariant decode( const QString & _b64, QVariant::Type _force_type )
 }
 
 
+//TODO C++20: It *may* be possible to make the following functions constepxr given C++20's constexpr std::string.
+
+/**
+ * @brief Base64 encodes data.
+ * @param data
+ * @return std::string containing the Base64 encoded data.
+ */
+std::string encode(std::string_view data)
+{
+	if (data.empty()) { return ""; }
+
+	// base64 encoded string
+	std::string result;
+
+	// number of chunks to process + padding
+	auto [numChunks, numTrailingBytes] = std::div(data.length(), numBytesPerChunk);
+
+	// add 1 chunk to handle last padded chunk
+	if (numTrailingBytes) { ++numChunks; }
+
+	// allocate one time to prevent in-loop reallocations
+	result.reserve(numChunks * numBase64CharPerChunk);
+
+	/*
+		char1               char2               char3
+		7 6 5 4 3 2   1 0 | 7 6 5 4   3 2 1 0 | 7 6   5 4 3 2 1 0
+		5 4 3 2 1 0 | 5 4   3 2 1 0 | 5 4 3 2   1 0 | 5 4 3 2 1 0
+		b64c1         b64c2           b64c3           b64c4
+	*/
+	// Get the first base64 character offset from the char chunk
+	auto b64char1 = [](std::string_view chunk) {
+		return static_cast<std::string::value_type>(chunk[0] >> 2);
+	};
+	// Get the second base64 character offset from the char chunk
+	auto b64char2 = [](std::string_view chunk) {
+		return static_cast<std::string::value_type>(
+			((chunk[0] & 0x03) << 4)
+			|
+			(chunk.size() > 0 ? ((chunk[1] & 0xF0) >> 4) : 0)
+		);
+	};
+	// Get the third base64 character offset from the char chunk
+	auto b64char3 = [](std::string_view chunk) {
+		return static_cast<std::string::value_type>(
+			(chunk.size() > 0 ? ((chunk[1] & 0x0F) << 2) : 0)
+			|
+			(chunk.size() > 1 ? ((chunk[2] & 0xC0) >> 6) : 0)
+		);
+	};
+	// Get the fourth base64 character offset from the char chunk
+	auto b64char4 = [](std::string_view chunk) {
+		return static_cast<std::string::value_type>(chunk.size() > 1 ? (chunk[2] & 0x3F) : 0);
+	};
+	for (int currentChunk = 0; currentChunk < numChunks; ++currentChunk)
+	{
+		std::string_view chunk = data.substr(currentChunk * numBytesPerChunk, numBytesPerChunk);
+		std::string output{
```

Comment:
Since this string is a fixed size of 4, you might get better performance in this loop by replacing `std::string` with `std::array` and using `result.append(output.data(), output.size());` down below.

Reply:
Changed to `std::array`. I think you're right, a fixed sized array could probably be more optimized than a `std::string`.